File: client/src/usb_detector/api_client.py
import json
import requests
import logging
from time import sleep
from diskcache import Deque
from requests import HTTPError, ConnectionError
from requests.exceptions import InvalidSchema


_uri = None     # server uri (url, port, and endpoint)
_cache = None   # cache (failed payloads)
_config = None  # instance of Config


def api_client_set_config(config):
    """Initializes the client API module.

    This function is meant to be called prior to calling any other function
    of the API module. It stores the instance of Config (config manager)
    into a private variable. It also initializes the cache for unsuccessful
    payloads and constructs a URI (endpoint on the server side).

    :param config: instance of Config which holds all values defined
                   in the configuration file.
    """
    # Store the variables globally within the module (file).
    global _config, _cache, _uri

    # Store the instance of Config and initialize the cache.
    _config = config
    _cache = _init_cache()

    # Creates the URI which is made of the server url, port, and path (endpoint).
    _uri = config.server_url + ":" + config.server_port + config.server_endpoint


def _init_cache():
    """ Initializes and returns a disk-based cache.

    The cache holds payloads that the application failed
    to send to the server. It periodically attempts to resent
    them to the server. All parameters can be seen in the
    configuration file.

    :return: instance of a new cache (Deque - FIFO)
    """
    return Deque(directory=_config.cache_dir)


def send_data(payload: dict):
    """Sends a payload off to the server.

    This function is called whenever a USB is connected
    or disconnected. If there is no internet connection or the
    server is not up and running, the payload will be stored
    into the disk cache.

    :param payload:
    """
    # Make sure that the URI has been constructed properly.
    # It's supposed to be done by calling the api_client_set_config function
    # with appropriate parameters.
    if _uri is None:
        logging.warning(f"sending payload = {payload} failed because uri is set to None")
        _cache_failed_payload(payload)
        return
    try:
        logging.info(f"sending payload = {payload} to {_uri}")
        response = requests.post(url=_uri, data=json.dumps(payload))
        logging.info(f"response text: {response.text}")
    except (ConnectionError, InvalidSchema):
        logging.warning(f"sending payload = {payload} to {_uri} failed")
        _cache_failed_payload(payload)
    except HTTPError as error:
        logging.error(f"HTTP Error ({_uri}) payload = {payload}, {error}")
        _cache_failed_payload(payload)


def _cache_failed_payload(payload: dict):
    """ Caches a payload.

    This function is called when the application fails to send a payload
    to the server. The payload gets stored into a file-based cache from which
    it will be periodically retrieved as the client will attempt to send
    it to the server again. All parameters regarding the cache can be found
    in the configuration file.

    :param payload: payload to be cached
    """
    # If the cache is "full", discard the oldest record.
    if len(_cache) >= _config.cache_max_entries:
        oldest_payload = _cache.pop()
        logging.warning(f"cache is full - discarding payload = {oldest_payload}")

    # Store the payload into the cache.
    logging.info(f"adding payload = {payload} into cache")
    _cache.append(payload)


def _resend_cached_payloads():
    """Reattempts to send cached payloads to the server (API).

    In the configuration file, there is a predefined number of
    payloads that can be sent to the server with each call of this function.
    This function is called periodically from api_client_run in order
    to resend failed payloads to the server.

    """
    # Calculate how many payload will be sent to the server
    retries = min(_config.cache_max_retries, len(_cache))
    logging.info(f"emptying the cache ({retries} records)")

    # Send the payloads to the server one by one.
    for _ in range(0, retries):
        payload = _cache.pop()
        send_data(payload)


def api_client_run():
    """ Keeps resending failed payloads to the server.

    This function is instantiated as a thread that periodically
    calls the _resend_cached_payloads function in order to empty
    the cache (failed payloads). The period can be set in the
    configuration file.
    """
    while True:
        # Resend a predefined amount of failed payloads to the server.
        _resend_cached_payloads()

        # Sleep for a predefined amount of seconds.
        sleep(_config.cache_retry_period_seconds)

File: ld_client/LDClient/network/ApiClient.cs
using System.Diagnostics;
using System.Net.Http.Json;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using DiskQueue;
using LDClient.network.data;

namespace LDClient.network {
    
    /// <summary>
    /// This class implements IApiClient which is an interface
    /// defining all the functionality required from an API client.
    /// </summary>
    public sealed class ApiClient : IApiClient {
        
        /// <summary>
        /// Instance of an HTTP client the is used to send data off to the server 
        /// </summary>
        public IHttpClient _client;

        /// <summary>
        /// Flag used to stop the client (periodical retrieval from the cache)
        /// </summary>
        public bool ClientRunning;

        /// <summary>
        /// Number of milliseconds after which the class tries to resend failed payloads to the server.
        /// </summary>
        private readonly uint _retryPeriod;
        
        /// <summary>
        /// Maximum number of entries (payloads) that can be sent to the server within one period (_retryPeriod).
        /// </summary>
        private readonly uint _maxEntries;
        
        /// <summary>
        /// Maximum number of failed payloads to be kept in the file-based cache (FIFO - when the maximum number is reached)
        /// </summary>
        private readonly uint _maxRetries;
        private readonly IPersistentQueue _cache;

        /// <summary>
        /// Creates an instance of the class.
        /// </summary>
        /// <param name="url">IP address of the server (url in case a DNS server is being used)</param>
        /// <param name="port">port that the API is running on</param>
        /// <param name="path">path of the API e.g. /api/v1/lg-logs</param>
        /// <param name="retryPeriod">number of milliseconds after which the class tries to resend failed payloads to the server</param>
        /// <param name="maxEntries">maximum number of entries (payloads) that can be sent to the server within one period</param>
        /// <param name="maxRetries">maximum number of failed payloads to be kept in the file-based cache</param>
        /// <param name="cache">instance of a persistent cache for storing failed payloads</param>
        public ApiClient(string url, uint port, string path, uint retryPeriod, uint maxEntries, uint maxRetries, IPersistentQueue cache) {
            // Construct the entire path to the API.
            var uri = $"{url}:{port}{path}";
            
            // Store the values into class variables.
            _retryPeriod = retryPeriod;
            _maxEntries = maxEntries;
            _maxRetries = maxRetries;
            _cache = cache;

            // Create an instance of a HttpClient which takes care of
            // establishing a connection to the server;
            _client = new HttpClient(uri);
        }

        /// <summary>
        /// Sends a payload to the server (API).
        /// </summary>
        /// <param name="payload">instance of a payload to be sent off to the server</param>
        public async Task SendPayloadAsync(Payload payload) {
            Program.DefaultLogger.Debug("SendPayloadAsync called.");
            try {
                // Create an instance of Stopwatch (to measure how much
                // the action took).
                Stopwatch stopWatch = new();
                
                // Send the payload to the server.
                stopWatch.Start();
                var response = await _client.PostAsJsonAsync(payload);
                stopWatch.Stop();
                
                // Create a log message.
                CreateRequestLog(payload, response, stopWatch.ElapsedMilliseconds);
                
                // Make sure the request was successful.
                response.EnsureSuccessStatusCode();
            } catch (Exception e) {
                Program.DefaultLogger.Error($"Failed to send {payload} to the server. Due to: {e.Message}");
                CachePayload(payload);
            }
        }

        /// <summary>
        /// Creates a request log message.
        /// </summary>
        /// <param name="payload">payload involved in the process of sending data to the server</param>
        /// <param name="response">response from the server</param>
        /// <param name="durationMs">duration in milliseconds (how much time it took to send off the payload)</param>
        private static void CreateRequestLog(Payload payload, HttpResponseMessage response, long durationMs) {
            // Create the log message.
            var responseToLog = new {
                statusCode = response.StatusCode,
                content = response.Content,
                headers = response.Headers,
                errorMessage = response.RequestMessage,
            };
            
            // Log the message using the logger defined in Program (main class).
            Program.DefaultLogger.Info($"Request completed in {durationMs} ms,\n" +
                                       $"Request body: {payload},\n" +
                                       $"Response: {responseToLog}");
        }
        
        /// <summary>
        /// Resends unsuccessful payloads to the server.
        /// </summary>
        private async Task ResendPayloadsAsync() {
            // Calculate the maximum number of payloads to be sent to the server.
            var numberOfPayloadsToResend = Math.Min(_maxRetries, _cache.EstimatedCountOfItemsInQueue);
            
            // Create a list for those payloads
            var payloads = new List<Payload>();
            
            // Retrieve the payloads from the cache.
            if (numberOfPayloadsToResend > 0) {
                // Open up a session to the cache.
                using var session = _cache.OpenSession();
                
                // Pop out payloads, deserialize them, and store them into the list.
                for (var i = 0; i < numberOfPayloadsToResend; i++) {
                    var rawBytes = session.Dequeue();
                    var payload = JsonSerializer.Deserialize<Payload>(rawBytes);
                    if (payload is not null) {
                        payloads.Add(payload);
                    }
                }
                // Flush the changes.
                session.Flush();
            }
            
            // If there are some payloads to be resent to the server.
            if (payloads.Count > 0) {
                Program.DefaultLogger.Debug($"ResendPayloadAsync -> {payloads.Count} unsent payloads");
                var tasks = new List<Task>();
                
                // Create a separate task for each payload - resend them to the server.
                foreach (var payload in payloads) {
                    Program.DefaultLogger.Info($"Resending {payload}.");
                    tasks.Add(SendPayloadAsync(payload));
                }
                // Wait until all tasks are finished. 
                await Task.WhenAll(tasks);
            }
        }
        
        /// <summary>
        /// Stores a failed payload into a persistent cache.
        /// </summary>
        /// <param name="payload"></param>
        private void CachePayload(Payload payload) {
            Program.DefaultLogger.Info($"Storing {payload} into the cache.");
            
            // Number of payloads stored in the cache.
            var numberOfCachedPayloads = _cache.EstimatedCountOfItemsInQueue;
            
            // Open up a session to the cache.
            using var session = _cache.OpenSession();
            
            // If the cache is "full", make room for the latest failed
            // payload by discarding the oldest one.
            if (numberOfCachedPayloads >= _maxEntries) {
                session.Dequeue();
            }
            
            // Store the payload into the cache.
            var payloadJson = JsonSerializer.Serialize(payload);
            session.Enqueue(Encoding.UTF8.GetBytes(payloadJson));
            
            // Flush the changes.
            session.Flush();
        }

        /// <summary>
        /// Runs the periodical retrieval of failed payloads stored
        /// in a file-based cache. This method is instantiated as a thread.
        /// </summary>
        public async void Run() {
            Program.DefaultLogger.Info("Api Client thread has started");
            
            // Keep the thread running.
            ClientRunning = true;
            
            // Keep resending failed payloads to the server.
            while (ClientRunning) {
                await ResendPayloadsAsync();
                Thread.Sleep((int) _retryPeriod);
            }
        }
    }
}
